task_sdk/src/airflow/sdk/api/client.py

```diff
@@ -0,0 +1,215 @@
+    "TaskInstanceOperations",
+]
+
+
+def get_json_error(response: httpx.Response):
+    """Raise a ServerResponseError if we can extract error info from the error."""
+    err = ServerResponseError.from_response(response)
+    if err:
+        log.warning("Server error", detail=err.detail)
+        raise err
+
+
+def raise_on_4xx_5xx(response: httpx.Response):
+    return get_json_error(response) or response.raise_for_status()
+
+
+# Py 3.11+ version
+def raise_on_4xx_5xx_with_note(response: httpx.Response):
+    try:
+        return get_json_error(response) or response.raise_for_status()
+    except httpx.HTTPStatusError as e:
+        if TYPE_CHECKING:
+            assert hasattr(e, "add_note")
+        e.add_note(
+            f"Correlation-id={response.headers.get('correlation-id', None) or response.request.headers.get('correlation-id', 'no-correlction-id')}"
+        )
+        raise
+
+
+if hasattr(BaseException, "add_note"):
+    # Py 3.11+
+    raise_on_4xx_5xx = raise_on_4xx_5xx_with_note
+
+
+def add_correlation_id(request: httpx.Request):
+    request.headers["correlation-id"] = str(uuid7())
+
+
+class TaskInstanceOperations:
+    __slots__ = ("client",)
+
+    def __init__(self, client: Client):
+        self.client = client
+
+    def start(self, id: uuid.UUID, pid: int, when: datetime):
+        """Tell the API server that this TI has started running."""
+        body = TIEnterRunningPayload(pid=pid, hostname=get_hostname(), unixname=getuser(), start_date=when)
+
+        self.client.patch(f"task_instance/{id}/state", content=self.client.encoder.encode(body))
+
+    def finish(self, id: uuid.UUID, state: TaskInstanceState, when: datetime):
+        """Tell the API server that this TI has reached a terminal state."""
+        body = TITerminalStatePayload(end_date=when, state=TerminalState(state))
+
+        self.client.patch(f"task_instance/{id}/state", content=self.client.encoder.encode(body))
+
+    def heartbeat(self, id: uuid.UUID):
+        self.client.put(f"task_instance/{id}/heartbeat")
+
+
+class ConnectionOperations:
+    __slots__ = ("client", "decoder")
+
+    def __init__(self, client: Client):
+        self.client = client
+        self.decoder: msgspec.json.Decoder[ConnectionResponse] = msgspec.json.Decoder(type=ConnectionResponse)
+
+    def get(self, id: str) -> ConnectionResponse:
+        """Get a connection from the API server."""
+        resp = self.client.get(f"connection/{id}")
+        return self.decoder.decode(resp.read())
+
+
+class BearerAuth(httpx.Auth):
+    def __init__(self, token: str):
+        self.token: str = token
+
+    def auth_flow(self, request: httpx.Request):
+        if self.token:
+            request.headers["Authorization"] = "Bearer " + self.token
+        yield request
+
+
+def noop_handler(request: httpx.Request) -> httpx.Response:
+    log.debug("Dry-run request", method=request.method, path=request.url.path)
+    return httpx.Response(200, json={"text": "Hello, world!"})
```

Comment:
What is this for?

Reply:
`airflow task test` equivalent, and an aid to debug and implementation.

Reply:
Can you add a comment for this... nor other noops like me coming around or my future me? :-D